lib/improver/nowcasting/lightning.py

```diff
@@ -0,0 +1,468 @@
+        # check prob-ice threshold units are as expected
+        ice_cube.coord('threshold').convert_units('kg m^-2')
+        new_cube_list = iris.cube.CubeList([])
+        err_string = "No matching prob(Ice) cube for threshold {}"
+        for cube_slice in prob_lightning_cube.slices_over('time'):
+            fcmins = cube_slice.coord('forecast_period').points[0]
+            for threshold, prob_max in zip(self.ice_thresholds,
+                                           self.ice_scaling):
+                ice_slice = ice_cube.extract(
+                    iris.Constraint(
+                        threshold=lambda t: isclose(t.point, threshold)))
+                if not isinstance(ice_slice, iris.cube.Cube):
+                    raise ConstraintMismatchError(err_string.format(threshold))
+                # Linearly reduce impact of ice as fcmins increases to 2H30M.
+                ice_scaling = [0., (prob_max * (1. - (fcmins / 150.)))]
+                cube_slice.data = np.maximum(
+                    rescale(ice_slice.data,
+                            data_range=(0., 1.),
+                            scale_range=ice_scaling,
+                            clip=True),
+                    cube_slice.data)
+            new_cube_list.append(cube_slice)
+
+        new_cube = new_cube_list.merge_cube()
+        new_cube = check_cube_coordinates(
+            prob_lightning_cube, new_cube)
+        return new_cube
+
+    def process(self, cubelist):
+        """
+        Produce Nowcast of lightning probability
+
+        Args:
+            cubelist (iris.cube.CubeList):
+                Contains cubes of
+                    * First-guess lightning probability
+                    * Nowcast precipitation probability
+                        (threshold > 0.5, 7., 35.)
+                    * Nowcast lightning rate
+                    * (optional) Analysis of vertically integrated ice (VII)
+                      from radar thresholded into probability slices for
+                      0.5, 1.0, 2.0 kg m-2
+
+        Returns:
+            new_cube (iris.cube.Cube):
+                Output cube containing Nowcast lightning probability.
+                This cube will have the same dimensions as the input
+                Nowcast precipitation probability.
+
+        Raises:
+            iris.exceptions.ConstraintMismatchError:
+                If cubelist does not contain the expected cubes.
+        """
+        first_guess_lightning_cube = cubelist.extract(
+            "probability_of_lightning", strict=True)
+        lightning_rate_cube = cubelist.extract(
+            "rate_of_lightning", strict=True)
+        lightning_rate_cube.convert_units("min^-1")  # Ensure units are correct
+        prob_precip_cube = cubelist.extract(
+            "probability_of_precipitation", strict=True)
+        prob_vii_cube = cubelist.extract(
+            "probability_of_vertical_integral_of_ice")
+        if prob_vii_cube:
```

Comment:
The other cubes are extracted with the option "strict=True" which forces cubelist.extract to return a cube, not a cubelist. As this is an optional cube, it may not be present and "strict=True" would also cause an error to be raised. I could rewrite it to use a try-except block instead.

Reply:
Added comment.